### DISCUSSION TITLE
privateGPT installation guide on Debian 11

### DISCUSSION THREAD

**Discussion by raymond0517** [imartinez/privateGPT#759] [Q&A] [answered]
is there any link providing detail step by step guide to install privateGPT on debian 11?

Thank you

Raymond

**Comment by 000Shimmy** [comment, marked as answer]
# Debian 13 (testing) Install Notes
Here's a verbose copy of my install notes using the latest version of Debian 13 (Testing) a.k.a Trixie and the 6.5.x kernel. Also, keep in mind that PrivateGPT is cutting edge FOSS tech and it may not play nice with older systems; particularly older GPUs. That being said your mileage may vary depending on your hardware; so, per usual ___nvidia GPUs make you fight for the FOSS life___.

## Intro 
By default, Trixie doesn't trust snakes; she requires that anything Python be installed via `apt install python3-{name}`. However, some - _read "most"_ - of the python3 tools/apps/utils/etc are not available in the Debian repos. If you want to use any of those questionable snakes then they must be used within a pre-built virtual environment. There exists great arguments for and against this approach: I'll leave you to your opinions, and get on with the  _Debian way of installing PrivateGPT_. When done you should have a  PrivateGPT instance up and running on your machine.  Let's get to it then; shall we?

## _This is the_ Debian _Way_ 
```bash
sudo apt update 
sudo apt upgrade 
```

### Install PrivateGPT 
Change to the directory that you want to install the virtual python environment for PrivateGPT into.  In this example I will be using the Desktop directory, but you can use anyone  that you like. 
```bash
cd ~/Desktop 
git clone "https://github.com/imartinez/privateGPT.git"
```
To accommodate the Debian virtual environment requisite we have to deviate from the standard instructions just a bit. Don't change into the privateGPT directory just yet. First, as required by picky Trixie, you have to build and activate the virtual environment. Make sure that if you use GPG Wallet, such as KDE's kwallet; that you are signed into it or all the `poetry` dependencies install will fail. 

```bash
python3 -m venv ./privateGPT       # build the virtual environment 
source ./privateGPT/bin/activate   # activate the virtual environment 
```
You'll notice a change in your PS1 prompt. After you `cd` into the privateGPT directory you will be inside the virtual environment that you just built and activated for it. Use the `deactivate` command to shut it down. __NOTE__: with `entr` or another tool you can automate most activating and deactivating the virtual environment, along with starting the privateGPT server with a couple of scripts; even dmenu offers some viable actions, but that's off-topic though.

Some of the dependencies and language model files installed by `poetry` are quite large and depending upon your ISP's bandwidth speeds this part may take awhile. If you haven't had your coffee, or tea, warmed up in a while then immediately following this next set up commands will be the time to do accomplish that task. 

```bash
cd ./privateGPT  
pip install poetry                           # installs the version control installer 
poetry install --with ui                     # install dependencies
poetry run python scripts/setup              # installs models
```
When that's done you will have access to your own privateGPT available at `localhost:8001` or `127.0.0.1:8001` after you run the following command.

```bash
poetry run python -m private_gpt   # runs the privateGPT server
```
### Back-Grounding PrivateGPT
After you get privateGPT up and running, test it out with some documents. Then you can run it in the background with the following command. __NOTE__: stdout redirects to /dev/null so all your goodness in your output is sent into the unknown abyss. 

```bash
poetry run python -m private_gpt > /dev/null 2>&1 & # logs into the abyss & and regain use of the terminal 
```

I hope this helped and may [God](https://www.kingjamesbibleonline.org/John-3-16/) [bless you](https://www.kingjamesbibleonline.org/Numbers-6-24_6-26/).

**Reply by pierri1101** [reply]
_poetry install --with ui_ didn't work for me.

Instead I used: _poetry install --extras "ui"_

**Comment by janwarming** [comment]
Will it support more than one GPU, I have a mining rig I like to use for AI.

**Comment by 000Shimmy** [comment]
The python environment encapsulates the python operations of the privateGPT within the directory, but it’s not a container in the sense of podman or lxc. I don’t foresee any “breaking” issues assigning privateGPT more than one GPU from the OS as described in the docs. However, I don’t have any surplus GPUs at the moment to test this theory out, so…

If you can’t get it to work, post back and we will figure something else out. There’s more than one way to accomplish your goals.

**Reply by stonecan** [reply]
Right now, I can't get private-gpt working on Debian 12, with the most serious breaking issue being the inability to get nomic-embed-text-v1.5 working (keep getting huggingface errors even though everything should be working). Appreciate any help.

**Comment by reefreaper** [comment]
Thank you for this how to. I'm up and running with Kali latest and it works great. A little slow but I'm running on CPU.

**Comment by 0xj0n1** [comment]
```
  Login successful
(privateGPT) root@vs-16452 ~/privateGPT # huggingface-cli login

    _|    _|  _|    _|    _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|_|_|_|    _|_|      _|_|_|  _|_|_|_|
    _|    _|  _|    _|  _|        _|          _|    _|_|    _|  _|            _|        _|    _|  _|        _|
    _|_|_|_|  _|    _|  _|  _|_|  _|  _|_|    _|    _|  _|  _|  _|  _|_|      _|_|_|    _|_|_|_|  _|        _|_|_|
    _|    _|  _|    _|  _|    _|  _|    _|    _|    _|    _|_|  _|    _|      _|        _|    _|  _|        _|
    _|    _|    _|_|      _|_|_|    _|_|_|  _|_|_|  _|      _|    _|_|_|      _|        _|    _|    _|_|_|  _|_|_|_|

A token is already saved on your machine. Run `huggingface-cli whoami` to get more information or `huggingface-cli logout` if you want to log out.
Setting a new token will erase the existing one.
To log in, `huggingface_hub` requires a token generated from https://huggingface.co/settings/tokens .
Enter your token (input will not be visible):
Add token as git credential? (Y/n) n
Token is valid (permission: fineGrained).
Your token has been saved to /root/.cache/huggingface/token
Login successful
(privateGPT) root@vs-16452 ~/privateGPT # poetry run python scripts/setup
22:50:12.302 [INFO    ] private_gpt.settings.settings_loader - Starting application with profiles=['default']
Downloading embedding nomic-ai/nomic-embed-text-v1.5
Traceback (most recent call last):
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/utils/_http.py", line 406, in hf_raise_for_status
response.raise_for_status()
File "/root/privateGPT/lib/python3.11/site-packages/requests/models.py", line 1024, in raise_for_status
raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 401 Client Error: Unauthorized for url: https://huggingface.co/api/models/nomic-ai/nomic-embed-text-v1.5/revision/main

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/root/privateGPT/scripts/setup", line 23, in <module>
snapshot_download(
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
return fn(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/_snapshot_download.py", line 229, in snapshot_download
raise api_call_error
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/_snapshot_download.py", line 160, in snapshot_download
repo_info = api.repo_info(repo_id=repo_id, repo_type=repo_type, revision=revision, token=token)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
return fn(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/hf_api.py", line 2682, in repo_info
return method(
^^^^^^^
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
return fn(*args, **kwargs)
^^^^^^^^^^^^^^^^^^^
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/hf_api.py", line 2467, in model_info
hf_raise_for_status(r)
File "/root/privateGPT/lib/python3.11/site-packages/huggingface_hub/utils/_http.py", line 454, in hf_raise_for_status
raise _format(RepositoryNotFoundError, message, response) from e
huggingface_hub.errors.RepositoryNotFoundError: 401 Client Error. (Request ID: Root=1-67b50094-48a3e3c15018786344390f97;5e1db945-dbf7-434c-9c0a-fdb9add5b731)

Repository Not Found for url: https://huggingface.co/api/models/nomic-ai/nomic-embed-text-v1.5/revision/main.
Please make sure you specified the correct `repo_id` and `repo_type`.
If you are trying to access a private or gated repo, make sure you are authenticated.
Invalid credentials in Authorization header
```

Does anyone face similar issues?

**Reply by stonecan** [reply]
Yes, I have the exact same issue. But unfortunately I have not found a solution so far. My best guess is that something changed on the nomic-embed-text-v1.5 repository, but I have no idea what.

**Reply by Myria-de** [reply]
You need to set the HF_TOKEN (https://huggingface.co/settings/tokens). Then run
HF_TOKEN=[MyToken] poetry run python scripts/setup
(Reference: https://docs.privategpt.dev/quickstart/getting-started/quickstart#fully-local-setups)

**Reply by stonecan** [reply]
Thanks for your answer. I did set the HF_TOKEN env variable with the Huggingface token and still got the same error as @0xj0n1...

**Reply by Myria-de** [reply]
You can check the Huggingface login in your environment with
```huggingface-cli login --token [MyToken]```
and after a successful login the user name with
```huggingface-cli whoami```
If this works, the script should also work.
